spanconfigkvsubscriber: conditionally ignore ProtectionPolicies

adityamaru opened the pull request:
This change teaches the `GetProtectionTimestamps` method in the KVSubscriber
to ignore ProtectionPolicies that were written by a backup, and apply to a span
that has been marked as excluded from backup. This ensures that the ProtectionPolicy
written by a backup does not holdup GC on the span since it will not be exporting its
row data.

Informs: https://github.com/cockroachdb/cockroach/issues/73536

Release note: None

Release justification: low risk update to new functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77392)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
Change looks good as is, but IMO an alternative worth considering would be to keep the interface as is and teach the KVSubscriber's implementation to ignore certain protection policies when querying it for protected timestamps. The KVSubscriber already has the SpanConfig and this would ensure the caller doesn't have to join the two fields to filter out certain protection policies. 

> Reviewed 12 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-:-MxRNqnHBmoWmNZx9Lif:b-mp87tc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @arulajmani)

___
*[pkg/kv/kvserver/protectedts/ptcache/cache.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-MxRDJSfBK-B5Zgu_O-F:-MxRDJSfBK-B5Zgu_O-G:b8cqse) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb86f7156aeef6760048aaf3ee11a9ae5d66f7fd/pkg/kv/kvserver/protectedts/ptcache/cache.go#L139)):*
> ```Go
> 		func(rec *ptpb.Record) (wantMore bool) {
> 			protectionPolicies = append(protectionPolicies,
> 				roachpb.ProtectionPolicy{ProtectedTimestamp: rec.Timestamp})
> ```

Could you write a comment here explaining why the default value of `IgnoreIfExcludedFromBackup` is the right thing?
___
*[pkg/spanconfig/spanconfig.go, line 410 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-MxRDg2r2sEV_T0_UCQV:-MxRDg2r2sEV_T0_UCQW:b-wqak7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb86f7156aeef6760048aaf3ee11a9ae5d66f7fd/pkg/spanconfig/spanconfig.go#L410)):*
> ```Go
> // protectedts.Cache.
> type ProtectedTSReader interface {
> 	// GetProtectionPolicies returns all protected timestamps that apply to any
> ```

nit: s/all protected timestamps/all protection policies/g
___
*[pkg/kv/kvserver/replica_protected_timestamp_test.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-MxRJKV82tVU5tPKesA6:-MxRJKV82tVU5tPKesA7:b-nj69us) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb86f7156aeef6760048aaf3ee11a9ae5d66f7fd/pkg/kv/kvserver/replica_protected_timestamp_test.go#L63)):*
> ```Go
> 				mp.protections = append(mp.protections, manualPTSReaderProtection{
> 					sp:                 roachpb.Span{Key: keys.MinKey, EndKey: keys.MaxKey},
> 					protectionPolicies: []roachpb.ProtectionPolicy{{ProtectedTimestamp: ts}},
> ```

nit: for all these, you could define yourself a helper that constructs protection policies given a list of timestamps. 
___
*[pkg/kv/kvserver/protectedts/ptcache/cache_test.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-MxRN8ib5m1fPxOdnUTJ:-MxRN8ib5m1fPxOdnUTK:be6ds6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb86f7156aeef6760048aaf3ee11a9ae5d66f7fd/pkg/kv/kvserver/protectedts/ptcache/cache_test.go#L412)):*
> ```Go
> 						protections[j].ProtectedTimestamp)
> 				})
> 				require.Equal(t, []roachpb.ProtectionPolicy{
> ```

nit: Similar to my comment above, might be worth extracting this out in a helper that can give you a list of protection policies given a list of timestamps. 
___
*[pkg/spanconfig/spanconfigptsreader/adapter_test.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-MxRNLn7-c5m41T6XriT:-MxRNLn7-c5m41T6XriU:b-8htneg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb86f7156aeef6760048aaf3ee11a9ae5d66f7fd/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L68)):*
> ```Go
> 
> 	// Add some records to the cache; ensure they're returned.
> 	mc.protectionPolicies = append(mc.protectionPolicies,
> ```

nit: here as well, if you decide to take my suggestion. 


<!-- Sent from Reviewable.io -->

adityamaru committed:
spanconfigkvsubscriber: conditionally ignore ProtectionPolicies

This change teaches the `GetProtectionTimestamps` method in the KVSubscriber
to ignore ProtectionPolicies that were written by a backup, and apply to a span
that has been marked as excluded from backup. This ensures that the ProtectionPolicy
written by a backup does not holdup GC on the span since it will not be exporting its
row data.

Informs: #73536

Release note: None

Release justification: low risk update to new functionality

arulajmani reviewed (approved):
Thanks for making the change, this is much cleaner! 

> Reviewed 14 of 16 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-:-Mxa27pQ9UAcS7FOyuVN:b-lc26q3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @arulajmani)

___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-Mxa25KJ20_shsborloJ:-Mxa25KJ20_shsborloK:b44jxcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06d5dfe7f4e18543283e97e8e5755c2b1a926e0/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go#L33)):*
> ```Go
> }
> 
> func tableSpan(tableID uint32) roachpb.Span {
> ```

nit: can we make this a local function to the test? 
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-Mxa1hEK51YZSoFgB45W:-Mxa1hEK51YZSoFgB45X:bnkcbt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06d5dfe7f4e18543283e97e8e5755c2b1a926e0/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go#L65)):*
> ```Go
> 	sp43Cfg.cfg.ExcludeDataFromBackup = true
> 
> 	subscriber := &KVSubscriber{}
> ```

Can we create a test cluster to get all dependencies and replace this with a call to `spanconfigkvsubscriber.New` instead?
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go, line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-Mxa2-YZ492P7axwdEu8:-Mxa2-YZ492P7axwdEu9:b44jxcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06d5dfe7f4e18543283e97e8e5755c2b1a926e0/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go#L122)):*
> ```Go
> }
> 
> func makeSpanAndSpanConfigWithProtectionPolicies(
> ```

nit: can we make this a local function to the test? 
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-Mxa1usU47rG6BqSMOTM:-Mxa1usU47rG6BqSMOTN:b-qnebqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06d5dfe7f4e18543283e97e8e5755c2b1a926e0/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go#L143)):*
> ```Go
> 	context.Context, bool, ...spanconfig.Update,
> ) (deleted []spanconfig.Target, added []spanconfig.Record) {
> 	return nil, nil
> ```

Here, and below, let's make these panic? 


<!-- Sent from Reviewable.io -->

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77392)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go, line 65 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-Mxa1hEK51YZSoFgB45W:-Mxbvpuc9WDeHJPUGvHN:b-1nxdfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06d5dfe7f4e18543283e97e8e5755c2b1a926e0/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go#L65)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we create a test cluster to get all dependencies and replace this with a call to `spanconfigkvsubscriber.New` instead?
</blockquote></details>

just called into `spanconfigkvsubscriber.New` with everything nil'ed out and that works too.
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77392#-Mxa1usU47rG6BqSMOTM:-Mxbs41MCeK3Fi7hTE10:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06d5dfe7f4e18543283e97e8e5755c2b1a926e0/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber_test.go#L143)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Here, and below, let's make these panic? 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adityamaru commented:
pretty sure `"show cluster setting version" timed out after 2m0s (took 2m0.001s): value differs between local setting ([18 8 8 21 16 2 24 0 32 0]) and KV ([18 9 8 21 16 1 24 0 32 144 9]); try again later` is a flake. Thanks for the review on this!

bors r=arulajmani
